# Description
The MDEHT package executes multivariate Hotelling’s T-square test to detect differential expression in isomiRs RNA-Seq data. Hotelling’s T-square test is a generalization of Student's t- statistic and is widely used for testing the difference between two multivariate means. Currently, the MDEHT package performs both one and two samples multivariate test. In one sample test, the null hypothesis is defined as the vector of the mean value of multiple variables equal to zero. 


## Access
MDEHT and isomiRseeker are free software tools that can be used by academic groups for non-commercial purpose. 

## System requirements  
> Linux/Unix or Mac OS platform   
> R (https://www.r-project.org/)  

isomiRseeker requires the following programs to be installed in user’s environment.  
> samtools (http://www.htslib.org/download/)  
> bedtools (https://bedtools.readthedocs.io/en/stable/content/installation.html)  
> bwa (https://github.com/lh3/bwa)  
> Perl 5 (http://www.perl.org/) 

## Usage isomiRseeker 
To generate expression data of isomiR from read alignments in the format of bam files, run the following code:

> perl /path/isomiRseeker.pl -b -p output_file_name input_file.bam

**Note:** The output from the above code will produce two files for each sample on your working directory, one is read counts and another one is rpm. Users need to combine read counts or rpm of all samples into a matrix before running MDETH. 

### Example
> perl /path/isomiRseeker.pl -b -p TCGA-2F-A9KQ-01A-11R-A38M-13_mirna TCGA-2F-A9KQ-01A-11R-A38M-13_mirna.bam



## Usage MDEHT
The input file for the MDEHT is isomiR read count data in the format of matrix where rows represent isomiRs and columns represent samples. The row names of the data matrix are isomiR names that should be same as isomiR symbols generated by the isomiRseeker tool. We recommend that isomiR data are normalized by the count per million (CPM) normalization method before running MDEHT method.  


### Example
#required the following packages
> install.packages("devtools")  
> library(devtools) 
> install_github("amanzju/MDEHT") 
> library(edgeR)  
> library (MDEHT) 

#load the example dataset 
> data("ExampleData")

#use CPM normalization from edgeR package 
> NormalizedData = log2(cpm(ExampleData)+1) 

#separate sample group  
> TumorData = NormalizedData[,grep("01A", colnames(NormalizedData))]  
> NormalData = NormalizedData[,grep("11A", colnames(NormalizedData))] 

#view the summary of isomiRs and UTR in plot  
> isomiRs<-rownames(TumorData)  
> isomiR.summary(x = isomiRs) 
> utr.summary(x = isomiRs)  

> results = MDEHT (X = TumorData, Y = NormalData) 
> head(results) 

## Output
The output of MDEHT is a summary of putative miRNAs, including the mean values for two groups, the value of the T-statistic, p-value and FDR adjusted p-value and log2 fold change value. Similarly, for one sample test, the output will provide the value of the T-statistic, p-value, and FDR adjusted p-value.


## Restriction of MDEHT
MDEHT will stop when the number of isoforms in miRNA is larger than n1 + n2 – 2, where n1 and n2 are sample sizes of two different groups. This restriction can be released by using
the technical parameters inohf.isomiRs and merge.olf.isomRs in the MDEHT function. When a miRNA contains a large number of isomiRs, the use of inohf.isomiRs will prompt the MDEHT to analyse only high-frequency isoforms, while the use of merge.olf.isomRs will prompt the MDEHT to merge low-frequency isoforms into one group. 
